Minor fix record collector

mjsax opened the pull request:
If `partition==null` and `partitioner!=null` we should not fall back to default partitioner (as we do before the patch if `producer.partitionsFor(...)` returns empty list. Falling back to default partitioner might corrupt hash partitioning.

mjsax committed:
Fix RecordCollector

mjsax committed:
code cleanup

mjsax commented:
Call for review @enothereska @dguy (note, it's two commits: (1) for the fix, (2) for code cleanup only)

I also changed the interface, as it does not make sense to provide a `partition` and a custom `partitioner` at the same time.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:
```diff
@@ -46,54 +46,60 @@
+        final ProducerRecord<byte[], byte[]> serializedRecord =
                 new ProducerRecord<>(topic, partition, timestamp, keyBytes, valBytes);
 
         for (int attempt = 1; attempt <= MAX_SEND_ATTEMPTS; attempt++) {
```
Nit: can we start at attempt = 0 and attempt < MAX_SEND_ATTEMPTS for consistency?

mjsax replied:
I only added a comment -- IMHO, it's better to start counting from 1 for this specific case.

enothereska reviewed (commented):
Is it worth adding a test to catch the exception thrown?

enothereska commented:
Thanks for separating fix from cleanup, was helpful.

mjsax committed:
Github comments

mjsax commented:
Updated: Added a test.

> Thanks for separating fix from cleanup, was helpful.

Lesson learned. It's not so hard, if you consider this from the beginning on :)

dguy reviewed (approved):
LGTM

mjsax commented:
Call for final review and merging @guozhangwang

guozhangwang commented:
retest this please

guozhangwang reviewed (approved):
Merged to trunk.
